File: drone_gcs/python_service/observability/health.py
"""Liveness/readiness state for the Python MAVLink service (Phase 5A).

Liveness = "the event loop is alive and we are not shutting down" → a failing
/healthz should make a supervisor restart the process. Readiness = "safe to
receive traffic": the telemetry publisher must be up and (after a cold-start grace
window) the publish loop must be ticking. This lets an orchestrator drain a worker
whose telemetry loop has wedged even while the HTTP server still answers.
"""
from __future__ import annotations

import time
from typing import Callable, Dict, Optional


class HealthState:
    def __init__(self, *, grace_s: float = 20.0, tick_stale_s: float = 5.0,
                 now: Optional[Callable[[], float]] = None) -> None:
        self._now = now or time.time
        self.grace_s = grace_s
        self.tick_stale_s = tick_stale_s
        self.started_at = self._now()
        self.shutting_down = False
        self.publisher_up = False
        self.last_tick_at: Optional[float] = None

    def reset_for_startup(self) -> None:
        """Clear shutdown/tick state and restart the grace window. Called at each
        lifespan startup so a re-init (e.g. a supervised restart in-process, or a
        test harness reusing the app) begins from a clean, not-shutting-down state."""
        self.started_at = self._now()
        self.shutting_down = False
        self.last_tick_at = None

    def mark_shutting_down(self) -> None:
        self.shutting_down = True

    def set_publisher_up(self, up: bool) -> None:
        self.publisher_up = bool(up)

    def record_tick(self, at: Optional[float] = None) -> None:
        self.last_tick_at = at if at is not None else self._now()

    def liveness(self) -> Dict[str, object]:
        return {
            "status": "shutting_down" if self.shutting_down else "ok",
            "uptime_s": round(self._now() - self.started_at, 1),
        }

    def readiness(self) -> Dict[str, object]:
        now = self._now()
        within_grace = (now - self.started_at) < self.grace_s
        tick_age = None if self.last_tick_at is None else now - self.last_tick_at
        tick_fresh = tick_age is not None and tick_age <= self.tick_stale_s

        checks = {
            "publisher_up": self.publisher_up,
            "tick_fresh": tick_fresh,
            "last_tick_age_s": None if tick_age is None else round(tick_age, 1),
            "within_startup_grace": within_grace,
            "shutting_down": self.shutting_down,
        }

        if self.shutting_down:
            ready = False
        elif within_grace:
            ready = self.publisher_up
        else:
            ready = self.publisher_up and tick_fresh

        return {"ready": ready, "status": "ready" if ready else "not_ready", "checks": checks}

File: drone_gcs/node_api/lib/health.js
/**
 * Liveness/readiness state for the Node API gateway (Phase 5A).
 *
 * Liveness answers "is the event loop alive and not shutting down" — a failing
 * /healthz should trigger a supervisor restart. Readiness answers "should this
 * instance receive traffic" — it goes degraded when the telemetry stream is
 * silent past the stale threshold (outside the cold-start grace window), so a
 * load balancer / orchestrator stops routing to a gateway whose upstream data
 * has dried up while /healthz stays green.
 */
'use strict';

class HealthState {
  /**
   * @param {{ now?: () => number, staleMs: number, graceMs: number }} opts
   */
  constructor(opts) {
    this._now = opts.now || Date.now;
    this.staleMs = opts.staleMs;
    this.graceMs = opts.graceMs;
    this.startedAt = this._now();
    this.shuttingDown = false;
    this.zmqConnected = false;
    this.lastFrameAt = null; // ms epoch of last received ZMQ frame
  }

  markShuttingDown() {
    this.shuttingDown = true;
  }

  setZmqConnected(connected) {
    this.zmqConnected = !!connected;
  }

  recordFrame(atMs) {
    this.lastFrameAt = atMs != null ? atMs : this._now();
  }

  /** @returns {{ status: 'ok'|'shutting_down', uptime_s: number }} */
  liveness() {
    return {
      status: this.shuttingDown ? 'shutting_down' : 'ok',
      uptime_s: Math.round((this._now() - this.startedAt) / 1000),
    };
  }

  /**
   * @returns {{ ready: boolean, status: string, checks: object }}
   */
  readiness() {
    const now = this._now();
    const withinGrace = now - this.startedAt < this.graceMs;
    const frameAgeMs = this.lastFrameAt == null ? null : now - this.lastFrameAt;
    const telemetryFresh = frameAgeMs != null && frameAgeMs <= this.staleMs;

    const checks = {
      zmq_connected: this.zmqConnected,
      telemetry_fresh: telemetryFresh,
      last_frame_age_s: frameAgeMs == null ? null : Math.round(frameAgeMs / 1000),
      within_startup_grace: withinGrace,
      shutting_down: this.shuttingDown,
    };

    // Not ready while shutting down (drain), or when we've lost a previously
    // healthy telemetry stream. During the cold-start grace window an empty
    // stream is tolerated (nothing may be publishing yet).
    let ready;
    if (this.shuttingDown) {
      ready = false;
    } else if (withinGrace) {
      ready = this.zmqConnected;
    } else {
      ready = this.zmqConnected && telemetryFresh;
    }

    return { ready, status: ready ? 'ready' : 'not_ready', checks };
  }
}

module.exports = { HealthState };
